File: src/titan/ast/grammar/token/KeyWordDfaTokenAutomata.java
package titan.ast.grammar.token;

/**
 * 识别token的dfa,增加了keyword的判定，只保留依赖的数据，真正的实现在runtime里面.
 *
 * @author tian wei jun
 */
public class KeyWordDfaTokenAutomata extends DfaTokenAutomata {

  KeyWordAutomata keyWordAutomata;

  /**
   * 初始化oneTokenStringBuilder、eof、dfa.
   *
   * @param keyWordAutomata keyWord自动机
   * @param dfa 识别token的 确定有限状态自动机
   */
  public KeyWordDfaTokenAutomata(KeyWordAutomata keyWordAutomata, TokenDfa dfa) {
    super(dfa);
    this.keyWordAutomata = keyWordAutomata;
  }
}

File: runtime/rust/titan-ast-runtime/titan-ast-runtime-lib/src/key_word_dfa_token_automata.rs
use crate::{ast::Token, key_word_automata::KeyWordAutomata};

#[derive(Clone)]
pub(crate) struct KeyWordDfaTokenAutomata {
    pub(crate) key_word_automata: KeyWordAutomata,
}

impl KeyWordDfaTokenAutomata {
    pub(crate) fn build_token(&self, tokens: &mut Vec<Token>) {
        self.key_word_automata.build_token(tokens);
    }
}

impl Default for KeyWordDfaTokenAutomata {
    fn default() -> Self {
        Self {
            key_word_automata: Default::default(),
        }
    }
}
